### PR TITLE
Infer hole in type from pattern

### LINKED ISSUE #265: Type checking fails when the type of a pattern is not given by the signature
Example:
```
module Tmp;

inductive Bool {
  false : Bool;
  true : Bool;
};

not : _ → _;
not false ≔ true;
not true ≔ false;

end;
```

Ideally the inference algorithm should be able to infer the type of both holes.

### PR DESCRIPTION
**Pull request by janmasrovira**
Fixes anoma/juvix#1335

### CODE REVIEW

File: tests/negative/265/M.juvix [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  true : Bool;
+};
+
+inductive Pair (A : Type) (B : Type) {
+  mkPair : A → B → Pair A B;
+};
+
+f : _ → _;
+f (mkPair false true) ≔ true;
+f true ≔ false;
```
Comment (jonaprieto) [inline tests/negative/265/M.juvix:14]: One little thing. If I test this file, the error is fine. One gets as expected:
```
The constructor true has type:
  Bool
but is expected to have type:
  Pair Bool Bool
```
However, if the function changes a bit to be:

```agda
f : _ → _;
f true ≔ false;
f (mkPair false true) ≔ true;
```

We get an error message that is not ok. It should be "... has type: Pair _ _" or something. Tbh, Idk rn how to make it simpler.

```
The constructor mkPair has type:
  Pair
but is expected to have type:
  Bool
```

File: test/TypeCheck/Positive.hs [hunk 2]
```diff
@@ -87,6 +87,10 @@ tests =
       "Implicit arguments"
       "MicroJuvix"
       "Implicit.juvix",
+    PosTest
+      "Pattern match a hole type"
+      "265"
```
Comment (jonaprieto) [inline test/TypeCheck/Positive.hs:92]: Better if you name the folder as "issue265". The number alone doesn't mean much.

File: tests/negative/265/M.juvix (hunk 1 above)
Comment (janmasrovira) [reply tests/negative/265/M.juvix:14]: indeed. I was aware of this an I thought it was easy enough to understand, but on a second thought, it looks like a bug.
Maybe we should use a different error message when the arguments of the inductive type are irrelevant?
something like this:
```
The constructor mkPair belongs to the inductive type:
  Pair
but is expected to belong to the inductive type:
  Bool
```

Comment (jonaprieto) [reply tests/negative/265/M.juvix:14]: Exactly. Let's use this last suggestion. I'm fine with that for the moment. ;)

File: test/TypeCheck/Positive.hs (hunk 2 above)
Comment (jonaprieto) [reply test/TypeCheck/Positive.hs:92]: 0k

File: tests/negative/265/M.juvix (hunk 1 above)
Comment (janmasrovira) [reply tests/negative/265/M.juvix:14]: Implemented in bd363933f7f5b860d729f062a106992e525b0896

**Pull request closed without merging**